pkg/scrape/discovery/file/file.go

```diff
@@ -0,0 +1,418 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/config"
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/discovery"
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/discovery/targetgroup"
+	pmodel "github.com/pyroscope-io/pyroscope/pkg/scrape/model"
+	"github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+var (
+	fileSDScanDuration = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Name:       "pyroscop_sd_file_scan_duration_seconds",
+			Help:       "The duration of the File-SD scan in seconds.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		})
+	fileSDReadErrorsCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "pyroscop_sd_file_read_errors_total",
+			Help: "The number of File-SD read errors.",
+		})
+	fileSDTimeStamp = NewTimestampCollector()
+
+	patFileSDName = regexp.MustCompile(`^[^*]*(\*[^/]*)?\.(json|yml|yaml|JSON|YML|YAML)$`)
+
+	// DefaultSDConfig is the default file SD configuration.
+	DefaultSDConfig = SDConfig{
+		RefreshInterval: model.Duration(5 * time.Minute),
+	}
+)
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+	prometheus.MustRegister(fileSDScanDuration, fileSDReadErrorsCount, fileSDTimeStamp)
+}
+
+// SDConfig is the configuration for file based discovery.
+type SDConfig struct {
+	Files           []string       `yaml:"files"`
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
```

Comment:
This is something we may support in future but as of now pyroscope server uses **kebab-case** for yaml:
```suggestion
	RefreshInterval model.Duration `yaml:"refresh-interval,omitempty"`
```